File: packages/ledger/src/invoice-label.ts
const SHORT_DATE = new Intl.DateTimeFormat("pt-BR", {
  day: "2-digit",
  month: "2-digit",
  year: "2-digit",
  timeZone: "UTC",
});

export type FinancialDocumentKind =
  | "unknown"
  | "credit_card_invoice"
  | "bank_statement"
  | "invoice_nfe";

/**
 * Nome financeiro estável para uma fatura.
 *
 * O vencimento é a data que a pessoa reconhece primeiro; na falta dele, usamos
 * o fim do ciclo. O nome físico do PDF nunca participa do rótulo.
 */
export function formatInvoiceLabel({
  issuer,
  dueDate,
  periodEnd,
}: {
  issuer?: string | null;
  dueDate?: string | null;
  periodEnd?: string | null;
}): string {
  const origin = issuer?.trim() || "Fatura";
  const date = dueDate ?? periodEnd;
  if (date === null || date === undefined || !/^\d{4}-\d{2}-\d{2}$/.test(date)) {
    return origin;
  }
  return `${origin} ${SHORT_DATE.format(new Date(`${date}T12:00:00Z`))}`;
}

/** Nome estável para qualquer documento, sem chamar extrato de fatura. */
export function formatDocumentLabel({
  issuer,
  dueDate,
  periodEnd,
  documentKind = "unknown",
}: {
  issuer?: string | null;
  dueDate?: string | null;
  periodEnd?: string | null;
  documentKind?: FinancialDocumentKind | null;
}): string {
  const kindLabel =
    documentKind === "bank_statement"
      ? "Extrato"
      : documentKind === "invoice_nfe"
        ? "Nota fiscal"
        : documentKind === "credit_card_invoice"
          ? "Fatura"
          : "Documento";
  const origin = issuer?.trim() || kindLabel;
  const date = dueDate ?? periodEnd;
  if (date === null || date === undefined || !/^\d{4}-\d{2}-\d{2}$/.test(date)) {
    return issuer?.trim() ? `${origin} · ${kindLabel}` : origin;
  }
  return `${origin} · ${kindLabel} ${SHORT_DATE.format(new Date(`${date}T12:00:00Z`))}`;
}

const MONTH_LABEL = new Intl.DateTimeFormat("pt-BR", {
  month: "long",
  year: "numeric",
  timeZone: "UTC",
});

/**
 * `2026-06` como "junho de 2026".
 *
 * Mora aqui, junto do rótulo de fatura, porque a tela cruzada por operadora e a
 * conversa nomeiam os MESMOS meses: duas formatações independentes divergiriam
 * na primeira diferença de locale, e a pessoa leria "junho" num lugar e "Jun"
 * no outro para a mesma linha. Devolve o mês cru quando o formato não é
 * `YYYY-MM` — inventar um nome de mês seria pior que mostrar o código.
 */
export function formatMonthLabel(yearMonth: string): string {
  if (!/^\d{4}-\d{2}$/.test(yearMonth)) return yearMonth;
  return MONTH_LABEL.format(new Date(`${yearMonth}-01T12:00:00Z`));
}

File: services/agent/clara/ledger/invoice_label.py
"""Nome financeiro estável para um documento. Porta de `invoice-label.ts`.

O nome físico do PDF nunca participa do rótulo — é o que a interface mostra no
lugar de `filename`, que é só um identificador técnico interno.
"""

from __future__ import annotations

import re
from typing import Literal

FinancialDocumentKind = Literal["unknown", "credit_card_invoice", "bank_statement", "invoice_nfe"]

_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_MONTH_RE = re.compile(r"^\d{4}-\d{2}$")

_MONTH_NAMES_PT = (
    "janeiro", "fevereiro", "março", "abril", "maio", "junho",
    "julho", "agosto", "setembro", "outubro", "novembro", "dezembro",
)


def _short_date(iso_date: str) -> str:
    y, m, d = iso_date.split("-")
    return f"{d}/{m}/{y[2:]}"


def format_invoice_label(
    *, issuer: str | None = None, due_date: str | None = None, period_end: str | None = None
) -> str:
    """O vencimento é a data que a pessoa reconhece primeiro; na falta, o fim do ciclo."""
    origin = issuer.strip() if issuer and issuer.strip() else "Fatura"
    date = due_date or period_end
    if date is None or not _DATE_RE.match(date):
        return origin
    return f"{origin} {_short_date(date)}"


def format_document_label(
    *,
    issuer: str | None = None,
    due_date: str | None = None,
    period_end: str | None = None,
    document_kind: FinancialDocumentKind | None = "unknown",
) -> str:
    """Nome estável para qualquer documento, sem chamar extrato de fatura."""
    kind_label = {
        "bank_statement": "Extrato",
        "invoice_nfe": "Nota fiscal",
        "credit_card_invoice": "Fatura",
    }.get(document_kind or "unknown", "Documento")

    origin = issuer.strip() if issuer and issuer.strip() else kind_label
    date = due_date or period_end
    if date is None or not _DATE_RE.match(date):
        return f"{origin} · {kind_label}" if (issuer and issuer.strip()) else origin
    return f"{origin} · {kind_label} {_short_date(date)}"


def format_month_label(year_month: str) -> str:
    """`2026-06` como "junho de 2026". Devolve o mês cru se o formato não bater —
    inventar um nome de mês seria pior que mostrar o código."""
    if not _MONTH_RE.match(year_month):
        return year_month
    year, month = year_month.split("-")
    month_index = int(month) - 1
    if not (0 <= month_index < 12):
        return year_month
    return f"{_MONTH_NAMES_PT[month_index]} de {year}"
